### PR TITLE
[clang][RISCV] Extend intrinsic size check variable from 16 -> 32 bits. NFC

### PR DESCRIPTION
**Pull request by 4vtomat**
We currently have over 67000 intrinsics, uint16_t will overflow.

### CODE REVIEW

**Review by topperc — COMMENTED** [review]
Why doesn't any test catch it?

**Comment by jerryzj** [comment]
> Why doesn't any test catch it?

I suppose the [CI](https://github.com/riscv-non-isa/rvv-intrinsic-doc/actions/runs/11216383892/job/31175562939) of intrinsic doc should catch this, but it is not reporting errors as expected.
I happen to notice the failures and working on a fix now.

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Comment by topperc** [comment]
Why don’t any of our lit tests that use every intrinsic catch it?

File: clang/lib/Sema/SemaRISCV.cpp [hunk 1]
```diff
@@ -399,7 +399,7 @@ void RISCVIntrinsicManagerImpl::InitRVVIntrinsic(
                                      Record.HasFRMRoundModeOp);
 
   // Put into IntrinsicList.
-  uint16_t Index = IntrinsicList.size();
+  uint32_t Index = IntrinsicList.size();
```
Comment (topperc) [inline clang/lib/Sema/SemaRISCV.cpp:402]: What is the type of Intrinsics on this line further down

```
Intrinsics.insert({Name, Index});
```

**Comment by wangpc-pp** [comment]
> Why don’t any of our lit tests that use every intrinsic catch it?

We don't see any error, is it because your downstream added some intrinsics?

**Comment by topperc** [comment]
Does the number intrinsics vary with the number of extensions enabled? Maybe we don't have a test with all extensions?

**Comment by 4vtomat** [comment]
> Why don’t any of our lit tests that use every intrinsic catch it?

Somehow if we add more "target-feature" the intrinsics increase, I'm not sure why.
For example:
```
test.c

#include "riscv_vector.h"
  vfloat16mf4x7_t test_vlseg7e16ff_v_f16mf4x7(const _Float16 *base, size_t *new_vl, size_t vl) {
  return __riscv_vlseg7e16ff_v_f16mf4x7(base, new_vl, vl);
}
```
Compile with this command:
```
clang -cc1 -triple riscv64-- -S -target-feature +v -target-feature +zvfh -target-abi lp64d test.c
```
generate 61674 intrinsics.
However if we add `-target-feature +zvfbfmin`
```
clang -cc1 -triple riscv64-- -S -target-feature +v -target-feature +zvfh -target-feature +zvfbfmin -target-abi lp64d test.c
```
generate 64568 intrinsics

it's weird.

**Comment by topperc** [comment]
> > Why don’t any of our lit tests that use every intrinsic catch it?
> 
> Somehow if we add more "target-feature" the intrinsics increase, I'm not sure why. For example:
> 
> ```
> test.c
> 
> #include "riscv_vector.h"
>   vfloat16mf4x7_t test_vlseg7e16ff_v_f16mf4x7(const _Float16 *base, size_t *new_vl, size_t vl) {
>   return __riscv_vlseg7e16ff_v_f16mf4x7(base, new_vl, vl);
> }
> ```
> 
> Compile with this command:
> 
> ```
> clang -cc1 -triple riscv64-- -S -target-feature +v -target-feature +zvfh -target-abi lp64d test.c
> ```
> 
> generate 61674 intrinsics. However if we add `-target-feature +zvfbfmin`
> 
> ```
> clang -cc1 -triple riscv64-- -S -target-feature +v -target-feature +zvfh -target-feature +zvfbfmin -target-abi lp64d test.c
> ```
> 
> generate 64568 intrinsics
> 
> it's weird.

Why is it weird? Some intrinsics require specific features. If that feature isn't enabled, the intrinsic is skipped.

**Comment by 4vtomat** [comment]
> > > Why don’t any of our lit tests that use every intrinsic catch it?
> 
> > 
> 
> > Somehow if we add more "target-feature" the intrinsics increase, I'm not sure why. For example:
> 
> > 
> 
> > ```
> 
> > test.c
> 
> > 
> 
> > #include "riscv_vector.h"
> 
> >   vfloat16mf4x7_t test_vlseg7e16ff_v_f16mf4x7(const _Float16 *base, size_t *new_vl, size_t vl) {
> 
> >   return __riscv_vlseg7e16ff_v_f16mf4x7(base, new_vl, vl);
> 
> > }
> 
> > ```
> 
> > 
> 
> > Compile with this command:
> 
> > 
> 
> > ```
> 
> > clang -cc1 -triple riscv64-- -S -target-feature +v -target-feature +zvfh -target-abi lp64d test.c
> 
> > ```
> 
> > 
> 
> > generate 61674 intrinsics. However if we add `-target-feature +zvfbfmin`
> 
> > 
> 
> > ```
> 
> > clang -cc1 -triple riscv64-- -S -target-feature +v -target-feature +zvfh -target-feature +zvfbfmin -target-abi lp64d test.c
> 
> > ```
> 
> > 
> 
> > generate 64568 intrinsics
> 
> > 
> 
> > it's weird.
> 
> 
> 
> Why is it weird? Some intrinsics require specific features. If that feature isn't enabled, the intrinsic is skipped.

Oh I see, but if we extend the size, the memory usage will grow.

**Review by topperc — CHANGES_REQUESTED** [review]
I think we need to change

```
struct RVVOverloadIntrinsicDef {
  // Indexes of RISCVIntrinsicManagerImpl::IntrinsicList.
  SmallVector<uint16_t, 8> Indexes;
};
```
and

``` 
  // Mapping function name to index of IntrinsicList.
  StringMap<uint16_t> Intrinsics;
```

Those places were both changed by e0092eae431956a2fd17f7ea88e7ba26d5e44f7e in addition to the place you've currently changed.

**Comment by 4vtomat** [comment]
> I think we need to change
> 
> ```
> struct RVVOverloadIntrinsicDef {
>   // Indexes of RISCVIntrinsicManagerImpl::IntrinsicList.
>   SmallVector<uint16_t, 8> Indexes;
> };
> ```
> 
> and
> 
> ```
>   // Mapping function name to index of IntrinsicList.
>   StringMap<uint16_t> Intrinsics;
> ```
> 
> Those places were both changed by [e0092ea](https://github.com/llvm/llvm-project/commit/e0092eae431956a2fd17f7ea88e7ba26d5e44f7e) in addition to the place you've currently changed.

Oh, nice catch! I didn't even remember I've changed this before lol~

**Review by topperc — APPROVED** [review]
LGTM

**Comment by goldsteinn** [comment]
Can this get merged?

**Pull request merged**